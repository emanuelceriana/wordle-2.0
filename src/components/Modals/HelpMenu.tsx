import { useContext, ReactNode } from "react";
import { IAppContext, AppContext } from "../../context/AppContext";
import { Modal } from "./Modal";
import styles from "./HelpMenu.module.scss";

interface SimpleExplanationProps {
  children: ReactNode;
}

const SimpleExplanation = ({ children }: SimpleExplanationProps) => {
  return <div className={styles["helpMenu__explanation"]}>{children}</div>;
};

interface LetterExplanationProps {
  children: ReactNode;
  color?: "green" | "yellow" | "black";
  letter: string;
  hasOcurrences?: boolean;
  ocurrences?: number;
}

const LetterExplanation = ({
  children,
  color,
  letter,
  hasOcurrences = false,
  ocurrences,
}: LetterExplanationProps) => {
  return (
    <div className={styles["helpMenu__caseBlock"]}>
      {hasOcurrences ? (
        <div>
          <span>x{ocurrences}</span>
          {letter}
        </div>
      ) : (
        <div className={styles[`helpMenu__caseBlock-${color}`]}>{letter}</div>
      )}
      <div className={styles["helpMenu__caseBlock-explanation"]}>
        {children}
      </div>
    </div>
  );
};

export const HelpMenu = () => {
  const { isHelpMenuOpen, setIsHelpMenuOpen } =
    useContext<IAppContext>(AppContext);

  return (
    <Modal
      isOpen={isHelpMenuOpen}
      onCloseRequest={() => setIsHelpMenuOpen(false)}
      showCloseButton={true}
      title="Help"
    >
      <div className={styles.helpMenu}>
        <SimpleExplanation>
          To start playing you need just to write a word with the length of the
          squares availables, you can use the virtual keyboard if you want.
        </SimpleExplanation>
        <SimpleExplanation>
          <b>Easy Mode:</b> For each try you don't need to write word. A chain
          of characters is valid. It's easy and boring. ðŸ¥± <i>ASDASD</i> should
          work.
        </SimpleExplanation>
        <SimpleExplanation>
          <b>Hard Mode:</b> For each try you need to write a real word. It's
          harder and exciting. ðŸ”¥ Did you know that <i>QWERTY</i> is accepted by
          Merriam-Webster dictionary?
        </SimpleExplanation>
        <LetterExplanation color="green" letter="W">
          If green it means that the word has this letter and it is in the
          correct position.
        </LetterExplanation>
        <LetterExplanation color="yellow" letter="O">
          If yellow it means that the word has this letter but it is not in the
          correct position.
        </LetterExplanation>
        <LetterExplanation color="black" letter="R">
          If black it means that the word has not this letter.
        </LetterExplanation>
        <LetterExplanation hasOcurrences={true} ocurrences={2} letter="D">
          That x(X) means that the letter is present in the word X times.
        </LetterExplanation>
      </div>
    </Modal>
  );
};

export default HelpMenu;
