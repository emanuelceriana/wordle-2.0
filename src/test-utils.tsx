import React, { ReactNode } from "react";
import { render } from "@testing-library/react";
import { AppContext } from "./context/AppContext";

const MockProvider: React.FC<{
  children: ReactNode | ReactNode[];
  value: object;
}> = ({ children, value }) => (
  <AppContext.Provider value={value}>{children}</AppContext.Provider>
);

export function renderWithContext(element: React.ReactElement, value: object) {
  return render(<MockProvider value={value}>{element}</MockProvider>);
}

export function withContext(element: React.ReactElement, value: object) {
  return <MockProvider value={value}>{element}</MockProvider>;
}
