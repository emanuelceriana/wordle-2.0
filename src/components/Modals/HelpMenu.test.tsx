import { render, screen, fireEvent, act } from "@testing-library/react";
import { HelpMenu, SimpleExplanation, LetterExplanation } from "./HelpMenu";
import { renderWithContext } from "../../test-utils";

interface MockValueContext {
  isHelpMenuOpen: boolean;
  setIsHelpMenuOpen: () => void;
}

describe("<HelpMenu />", () => {
  const setIsHelpMenuOpenMock = jest.fn();

  let mockValueContext: MockValueContext = {
    isHelpMenuOpen: true,
    setIsHelpMenuOpen: setIsHelpMenuOpenMock,
  };

  it("should render without errors", () => {
    const { asFragment } = renderWithContext(<HelpMenu />, mockValueContext);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should be closed if helMenuOpen is false", () => {
    mockValueContext.isHelpMenuOpen = false;

    const { queryByTestId } = renderWithContext(<HelpMenu />, mockValueContext);

    const helpMenu = queryByTestId("helpMenu");

    expect(helpMenu).toBe(null);
  });
  it("should be open if helMenuOpen is true", () => {
    mockValueContext.isHelpMenuOpen = true;

    const { getByTestId } = renderWithContext(<HelpMenu />, mockValueContext);

    const helpMenu = getByTestId("helpMenu");

    expect(helpMenu).toBeDefined();
  });
  it("should call onCloseRequest when close button is clicked", () => {
    const { getByRole } = renderWithContext(<HelpMenu />, mockValueContext);

    const closeButton = getByRole("button");

    act(() => {
      fireEvent.click(closeButton);
    });

    expect(mockValueContext.setIsHelpMenuOpen).toHaveBeenCalledWith(false);
  });
});

describe("SimpleExplanation", () => {
  it("renders children", () => {
    render(<SimpleExplanation>Test Explanation</SimpleExplanation>);
    const explanationElement = screen.getByText("Test Explanation");
    expect(explanationElement).toBeInTheDocument();
  });
});

describe("LetterExplanation", () => {
  it("renders letter and children", () => {
    render(
      <LetterExplanation letter="A">Test Letter Explanation</LetterExplanation>
    );
    const letterElement = screen.getByText("A");
    const explanationElement = screen.getByText("Test Letter Explanation");

    expect(letterElement).toBeInTheDocument();
    expect(explanationElement).toBeInTheDocument();
  });

  it("applies the specified color", () => {
    render(
      <LetterExplanation color="green" letter="B">
        Test Color Explanation
      </LetterExplanation>
    );
    const letterElement = screen.getByText("B");

    expect(letterElement).toHaveClass("helpMenu__caseBlock-green");
  });

  it("renders ocurrences if specified", () => {
    render(
      <LetterExplanation hasOcurrences={true} ocurrences={3} letter="C">
        Test Occurrences Explanation
      </LetterExplanation>
    );
    const ocurrencesElement = screen.getByText("x3");

    expect(ocurrencesElement).toBeInTheDocument();
  });
});
