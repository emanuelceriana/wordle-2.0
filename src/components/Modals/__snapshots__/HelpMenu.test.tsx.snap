// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HelpMenu /> should render without errors 1`] = `
<DocumentFragment>
  <div
    class="modalBg"
  />
  <div
    class="modalBg__modal"
  >
    <div
      class="modalBg__modal-title"
    >
      Help
    </div>
    <button
      class="modalBg__modal-cButton"
    >
      X
    </button>
    <div
      class="modalBg__modal-body"
    >
      <div
        class="helpMenu"
        data-testid="helpMenu"
      >
        <div
          class="helpMenu__explanation"
        >
          To start playing you need just to write a word with the length of the squares availables, you can use the virtual keyboard if you want.
        </div>
        <div
          class="helpMenu__explanation"
        >
          <b>
            Easy Mode:
          </b>
           For each try you don't need to write word. A chain of characters is valid. It's easy and boring. ðŸ¥± 
          <i>
            ASDASD
          </i>
           should work.
        </div>
        <div
          class="helpMenu__explanation"
        >
          <b>
            Hard Mode:
          </b>
           For each try you need to write a real word. It's harder and exciting. ðŸ”¥ Did you know that 
          <i>
            QWERTY
          </i>
           is accepted by Merriam-Webster dictionary?
        </div>
        <div
          class="helpMenu__caseBlock"
        >
          <div
            class="helpMenu__caseBlock-green"
          >
            W
          </div>
          <div
            class="helpMenu__caseBlock-explanation"
          >
            If green it means that the word has this letter and it is in the correct position.
          </div>
        </div>
        <div
          class="helpMenu__caseBlock"
        >
          <div
            class="helpMenu__caseBlock-yellow"
          >
            O
          </div>
          <div
            class="helpMenu__caseBlock-explanation"
          >
            If yellow it means that the word has this letter but it is not in the correct position.
          </div>
        </div>
        <div
          class="helpMenu__caseBlock"
        >
          <div
            class="helpMenu__caseBlock-black"
          >
            R
          </div>
          <div
            class="helpMenu__caseBlock-explanation"
          >
            If black it means that the word has not this letter.
          </div>
        </div>
        <div
          class="helpMenu__caseBlock"
        >
          <div>
            <span>
              x2
            </span>
            D
          </div>
          <div
            class="helpMenu__caseBlock-explanation"
          >
            That x(X) means that the letter is present in the word X times.
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
